src/YouTubePlayer.ts

```diff
@@ -0,0 +1,56 @@
+import type { AdsConfig } from '@guardian/commercial-core';
+import { log } from '@guardian/libs';
+import { loadYouTubeAPI } from './loadYouTubeIframeApi';
+
+type EmbedConfig = {
+    embedConfig: {
+        relatedChannels: string[];
+        adsConfig: AdsConfig;
+    };
+};
+
+type PlayerOptions = YT.PlayerOptions & EmbedConfig;
+
+class YouTubePlayer {
+    playerPromise: Promise<YT.Player>;
+
+    constructor(id: string, playerOptions: PlayerOptions) {
+        this.playerPromise = this.setPlayer(id, playerOptions);
+    }
+
+    private async setPlayer(id: string, playerOptions: PlayerOptions) {
+        const YTAPI = await loadYouTubeAPI();
+        const playerPromise = new Promise<YT.Player>((resolve, reject) => {
+            try {
+                const player = new YTAPI.Player(id, playerOptions);
+                resolve(player);
+            } catch (e) {
+                this.logError(e as Error);
+                reject(e);
+            }
+        });
+        return playerPromise;
+    }
+
+    private logError(e: Error) {
+        log('dotcom', `YouTubePlayer failed to load: ${e}`);
+    }
+
+    getPlayerState(): Promise<YT.PlayerState | void> {
+        return this.playerPromise
+            .then((player) => {
+                return player.getPlayerState();
+            })
+            .catch(this.logError);
+    }
+
+    stopVideo(): Promise<void> {
+        return this.playerPromise
+            .then((player) => {
+                player.stopVideo();
+            })
+            .catch(this.logError);
+    }
```

Comment:
I know you already mentioned you might wanna explore removing the listeners (on component unmount) using `removeEventListener`. Was just wondering if you managed doing that?

Reply:
@arelra and I have discovered that `youtube-player` doesn't actually remove event listeners by calling `removeEventListeners` on the underlying player. We think it would be a good idea to add this feature in order to avoid memory leaks but will work on this as a follow up in another PR.

Reply:
Sounds good to me.

Reply:
Nice catch, yes sounds good to do it in upcoming PR 👍